### PR TITLE
Convert chains of ifs to equivalent switch statements

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
Plus a little renaming/clarification

### CODE REVIEW

**Review by katelynsills — APPROVED** [review]
This looks good to me, but I think @jfparadis originally wrote it, so let's get his input.

File: packages/cosmic-swingset/lib/ag-solo/vats/vat-wallet.js [hunk 1]
```diff
@@ -27,51 +27,44 @@ function build(E, D, _log) {
...
               data: pursesState,
             };
-          }
-          return {};
-        }
 
-        if (type === 'walletGetInbox') {
-          if (inboxState) {
+          case 'walletGetInbox':
+            if (!inboxState) return {};
```
Comment (jfparadis) [inline packages/cosmic-swingset/lib/ag-solo/vats/vat-wallet.js]: New logic? Why empty object? Why not a use a boolean as other cases?

**Review by jfparadis — APPROVED** [review]
I don't think that switch is a benefit and I would advise against this change, even more in this case for these 2 reasons:
- We don't have "a chain of its". The logic here is to return early once as soon as a condition is met. 
- Switch statement don't have blocks. So a declaration like `const result` in one branch is actually visible in all branches. This forces the developer to add a scope `{}` (like in your PR) or to be creative if the same variable name makes sense in more than one branch. 

A sequence of ifs guarantees complete isolation. And there is always the possibility of accidental fall-through. If linting was mandating scope and preventing fall-through to prevent these issue, I'd be more ok with it.

**Comment by Chris-Hibbert** [comment]
The main reason I converted to case/switch is that all the tests check the same variable (`type` in both files). If the intent is that exactly one branch is taken, then case/switch seems to express that more clearly. With case/switch, once you see that the top-level is switch(type), you only have to validate that each case exits to understand the structure. With a sequence of if statements, you can't be sure until you've checked every if to be sure that it's testing the same thing. If we decide not to change to case/switch, I'd recommend annotating both of these files to indicate that the structure is intended to implement an exclusive branch.

Returning early is as true of case/switch as of the sequence of ifs. Sorry if "chain of ifs" wasn't clear about that. 

It's true that switch statements don't natively have block structure, but as you noted this is easy to remedy when it's needed. If one accidentally tries to declare the same variable in more than one branch, it is caught quickly by the compiler.

**Comment by jfparadis** [comment]
Thanks @Chris-Hibbert, great comments. I'm fine with it, just wanting to provide a sounding board.

We're looking for clarity, and it's not a clear win:
- nesting increases with a switch statement. 
- exclusive branch is also not more obvious.

It's the famous illusion: do you see a vase, or do you see two faces. :)

Maybe the actual problem is somewhere else: there is just too much code in one function. Maybe we need one function for the branching logic, and one function for generating each answer.

```js
if (condition1) {
   return fn1(params)
}
if (condition2) {
   return fn2(params)
}
return false
```

```js
switch(variable) {
   case value1: return fn1(params);
   case value2: return fn2(params);
   default: return false;
}
```

Using that syntax, there is no ambiguity in either case.

One last point: a switch limits the conditions to equality. Something to keep in mind.

**Comment by erights** [comment]
> One last point: a switch limits the conditions to equality. Something to keep in mind.

Good point! And it is not `Object.is` equality but rather `===` so-called equality, which has two surprises:

```js
switch (NaN) { case NaN: return 1; default: return 2; }  // return 2
switch (-0) { case 0: return 1; case -0: return 2; default: return 3; }  // return 1
```

**Comment by erights** [comment]
The Jessie grammar only accepts switch statements when each body uses curlies and terminates in a terminating statement like `return`, `throw`, `break`, or `continue`. The switch statement in question would be rejected by the Jessie grammar because it ends with

```js
default:
  return false;
```
rather than
```js
default: {
  return false;
}
```

**Comment by Chris-Hibbert** [comment]
Thanks, @jfparadis and @erights 

Given the issues with switch in javascript (I didn't realize that NaN and -0 had weird behavior under ===. Are we safe from that if all the `case`s have string values?) I would be happier with the bodies turned into explicit function calls.

> One last point: a switch limits the conditions to equality. Something to keep in mind.

It's stronger than that. The advantage I'm relying on is that it limits the conditions to all be equality  with the same parameter, which is true of exactly the kind of code I want to make this transformation to. When the code's purpose is to select one branch based on a single incoming value, switch is clearer, even if javascript makes it not appropriate for all situations.

**Comment by erights** [comment]
> Are we safe from that if all the cases have string values?

Yes. Aside from NaN and -0, `===` is perfectly well behaved and identical to `Object.is`

**Pull request merged**